File: backend/lib/legacy/Inter-Letter Spacing Uniformity.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
import math


def compute_inter_letter_spacing_uniformity(image_path, debug=False):
    # Read the image
    img = cv2.imread(image_path)
    if img is None:
        return {"error": "Image not found"}

    # Convert to grayscale
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # Binarize the image
    _, binary = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)

    # Find contours (letters)
    contours, _ = cv2.findContours(binary, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    # Get bounding boxes for each contour
    boxes = [cv2.boundingRect(cnt) for cnt in contours]

    # Sort contours by leftmost x coordinate
    boxes.sort(key=lambda b: b[0])

    # Calculate gaps between adjacent letters
    gaps = []
    for i in range(len(boxes) - 1):
        x_current_right = boxes[i][0] + boxes[i][2]
        x_next = boxes[i + 1][0]
        gap = x_next - x_current_right
        if gap > 0:  # Only consider positive gaps
            gaps.append(gap)

    results = {}

    if gaps:
        # Calculate spacing metrics
        avg_gap = np.mean(gaps)
        gap_std = np.std(gaps)

        # Normalize by median letter width for scale invariance
        median_width = np.median([w for x, y, w, h in boxes])

        if median_width > 0:
            norm_avg_gap = avg_gap / median_width
            norm_gap_std = gap_std / median_width
        else:
            norm_avg_gap = avg_gap
            norm_gap_std = gap_std

        # Store results
        results = {
            "raw_avg_gap": float(avg_gap),
            "raw_gap_std": float(gap_std),
            "median_letter_width": float(median_width),
            "normalized_avg_gap": float(norm_avg_gap),
            "normalized_gap_std": float(norm_gap_std),
            "gap_count": len(gaps)
        }

        # Qualitative assessment
        if 0.1 <= norm_avg_gap <= 0.5 and norm_gap_std < 0.2:
            results["assessment"] = "Likely italic handwriting (moderate, consistent spacing)"
        elif norm_avg_gap < 0.1:
            results["assessment"] = "Likely cursive handwriting (minimal spacing)"
        elif norm_avg_gap > 0.5:
            results["assessment"] = "Likely printed handwriting (large spacing)"
        else:
            results["assessment"] = "Indeterminate style"

    # Debug visualization
    if debug and gaps:
        plt.figure(figsize=(10, 6))

        # Original image
        plt.subplot(2, 2, 1)
        plt.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
        plt.title("Original Image")

        # Binary image with contours
        plt.subplot(2, 2, 2)
        debug_img = cv2.cvtColor(binary.copy(), cv2.COLOR_GRAY2BGR)
        for box in boxes:
            x, y, w, h = box
            cv2.rectangle(debug_img, (x, y), (x + w, y + h), (0, 255, 0), 2)
        plt.imshow(cv2.cvtColor(debug_img, cv2.COLOR_BGR2RGB))
        plt.title("Letter Detection")

        # Histogram of gaps
        plt.subplot(2, 2, 3)
        plt.hist(gaps, bins=20)
        plt.axvline(avg_gap, color='r', linestyle='dashed', linewidth=2)
        plt.title(f"Gap Distribution (Mean={avg_gap:.2f})")

        # Normalized gaps
        if median_width > 0:
            plt.subplot(2, 2, 4)
            norm_gaps = [g / median_width for g in gaps]
            plt.hist(norm_gaps, bins=20)
            plt.axvline(norm_avg_gap, color='r', linestyle='dashed', linewidth=2)
            plt.title(f"Normalized Gaps (Mean={norm_avg_gap:.2f})")

        plt.tight_layout()
        plt.show()

    return results


if __name__ == "__main__":
    # Replace with your actual image file
    image_path = "atest/4.png"
    curviness_results = compute_inter_letter_spacing_uniformity(image_path, debug=True)
    print(curviness_results)

File: backend/lib/italic/inter_letter_spacing_uniformity.m
function results = computeInterLetterSpacingUniformity(imagePath, debug)
    if nargin < 2
        debug = false;
    end

    % Read the image
    img = imread(imagePath);
    if isempty(img)
        results = struct('error', 'Image not found');
        return;
    end

    % Convert to grayscale if necessary
    if size(img,3) == 3
        gray = rgb2gray(img);
    else
        gray = img;
    end

    % Binarize the image using Otsu's method and invert the result
    level = graythresh(gray);
    binary = imbinarize(gray, level);
    binary = ~binary; % Invert so letters become white

    % Find connected components and extract bounding boxes
    cc = bwconncomp(binary);
    stats = regionprops(cc, 'BoundingBox');
    if isempty(stats)
        results = struct('error', 'No letters found');
        return;
    end

    % Create a matrix of bounding boxes: each row is [x, y, width, height]
    boxes = zeros(length(stats), 4);
    for i = 1:length(stats)
        boxes(i,:) = stats(i).BoundingBox;
    end

    % Sort boxes by the x coordinate (leftmost)
    boxes = sortrows(boxes, 1);

    % Calculate gaps between adjacent letters
    gaps = [];
    for i = 1:size(boxes,1)-1
        x_current_right = boxes(i,1) + boxes(i,3);
        x_next = boxes(i+1,1);
        gap = x_next - x_current_right;
        if gap > 0
            gaps = [gaps, gap];
        end
    end

    results = struct();
    if ~isempty(gaps)
        % Compute gap statistics
        avg_gap = mean(gaps);
        gap_std = std(gaps);
        letter_widths = boxes(:,3);
        median_width = median(letter_widths);

        % Normalize by median letter width for scale invariance
        if median_width > 0
            norm_avg_gap = avg_gap / median_width;
            norm_gap_std = gap_std / median_width;
        else
            norm_avg_gap = avg_gap;
            norm_gap_std = gap_std;
        end

        % Store results
        results.raw_avg_gap = avg_gap;
        results.raw_gap_std = gap_std;
        results.median_letter_width = median_width;
        results.normalized_avg_gap = norm_avg_gap;
        results.normalized_gap_std = norm_gap_std;
        results.gap_count = length(gaps);

        % Qualitative assessment based on normalized gaps
        if norm_avg_gap >= 0.1 && norm_avg_gap <= 0.5 && norm_gap_std < 0.2
            results.assessment = 'Likely italic handwriting (moderate, consistent spacing)';
        elseif norm_avg_gap < 0.1
            results.assessment = 'Likely cursive handwriting (minimal spacing)';
        elseif norm_avg_gap > 0.5
            results.assessment = 'Likely printed handwriting (large spacing)';
        else
            results.assessment = 'Indeterminate style';
        end

        % Debug visualization if enabled
        if debug
            figure;

            % Subplot 1: Original Image
            subplot(2,2,1);
            imshow(img);
            title('Original Image');

            % Subplot 2: Binary Image with bounding boxes
            subplot(2,2,2);
            imshow(binary);
            hold on;
            for i = 1:size(boxes,1)
                rectangle('Position', boxes(i,:), 'EdgeColor', 'g', 'LineWidth', 2);
            end
            hold off;
            title('Letter Detection');

            % Subplot 3: Histogram of raw gaps
            subplot(2,2,3);
            histogram(gaps, 20);
            hold on;
            yl = ylim;
            plot([avg_gap avg_gap], yl, 'r--', 'LineWidth', 2);
            hold off;
            title(sprintf('Gap Distribution (Mean=%.2f)', avg_gap));

            % Subplot 4: Histogram of normalized gaps
            if median_width > 0
                norm_gaps = gaps / median_width;
                subplot(2,2,4);
                histogram(norm_gaps, 20);
                hold on;
                yl = ylim;
                plot([norm_avg_gap norm_avg_gap], yl, 'r--', 'LineWidth', 2);
                hold off;
                title(sprintf('Normalized Gaps (Mean=%.2f)', norm_avg_gap));
            end
        end
    else
        results.error = 'No gaps found';
    end
end


% Example usage
imagePath = '/Users/jameswong/PycharmProjects/NoteMercy_Extension/backend/atest/5.png';  % Update with your image file path
results = computeInterLetterSpacingUniformity(imagePath, true);
disp(results);
